dace/frontend/python/replacements.py

```diff
@@ -814,90 +872,122 @@ def _convert_type(dtype1, dtype2, operator) -> Tuple[dace.dtypes.typeclass]:
+        elif dtype in signed_types:
+            coarse_types.append(1) # signed integer, bool
+        else:
+            coarse_types.append(0)  # unsigned integer
+    
+    casting = [None] * len(arguments)
 
-    left_cast = None
-    right_cast = None
+    if len(arguments) == 1:  # Unary operators
 
-    if _is_op_arithmetic(operator):
+        if operator == 'USub' and coarse_types[0] == 0:
+            result_type = eval('dace.int{}'.format(4 * datatypes[0].bytes))
+        elif _is_op_boolean(operator):
+            result_type = dace.bool_
+        else:
+            result_type = datatypes[0]
 
-        # Float division between integers
-        if operator == 'Div' and max(type1, type2) < 2:
-            max_bytes = max(dtype1.bytes, dtype2.bytes)
-            if type1 == type2 and type1 == 0:  # Unsigned integers
-                result_type = eval('dace.uint{}'.format(4 * max_bytes))
+    elif len(arguments) == 2:  # Binary operators
+
+        type1 = coarse_types[0]
+        type2 = coarse_types[1]
+        dtype1 = datatypes[0]
+        dtype2 = datatypes[1]
+        left_cast = None
+        right_cast = None
+
+        if _is_op_arithmetic(operator):
+
+            # Float/True division between integers
+            if operator == 'Div' and max(type1, type2) < 2:
+                # TODO: Leaving this here in case we implement a C/C++ flag
+                # max_bytes = max(dtype1.bytes, dtype2.bytes)
+                # if type1 == type2 and type1 == 0:  # Unsigned integers
+                #     result_type = eval('dace.uint{}'.format(8 * max_bytes))
+                # else:
+                #     result_type = eval('dace.int{}'.format(8 * max_bytes))
+                result_type = dace.float64
+                left_cast = dace.float64
+                right_cast = dace.float64
+            # Floor division with at least one complex argument
+            elif operator == 'FloorDiv' and max(type1, type2) == 3:
+                raise TypeError("can't take floor of complex number")
+            # Floor division with at least one float argument
+            elif operator == 'FloorDiv' and max(type1, type2) == 2:
+                result_type = dace.float64
+            # Floor division between integers
+            elif operator == 'FloorDiv' and max(type1, type2) < 2:
+                max_bytes = max(dtype1.bytes, dtype2.bytes)
+                if type1 == type2 and type1 == 0:  # Unsigned integers
+                    result_type = eval('dace.uint{}'.format(8 * max_bytes))
+                else:
+                    result_type = eval('dace.int{}'.format(8 * max_bytes))
+                right_cast = dace.float64
+            # Power with base integer and exponent signed integer
+            elif (operator == 'Pow' and max(type1, type2) < 2 and
+                    dtype2 in signed_types):
+                result_type = dace.float64
+                left_cast = dace.float64
+                right_cast = dace.float64
```

Comment:
Anything (float,complex) with integer exponent can keep the original type, right? I don't want float32 ** int32 to be float64

Reply:
Yes, this should trigger only when both arguments are integers. It also looks like NumPy sets the result to float64 regardless of the arguments being int16 or int32. Make sense, because when converting to float you care about the resolution.